### ISSUE TITLE
Epic: Extend trainer benchmark coverage to DQN and SAC (all four single-agent algos comparable)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#159]
### Goal
All four single-agent algorithms — **PPO, A2C, DQN, SAC** — should have comparable, reproducible benchmarks. Today only the two on-policy trainers (A2C, PPO) are covered by the throughput harness and the learning-curve CSV instrumentation. This epic closes the gap so we can honestly compare per-update cost and sample-efficiency across the full algorithm set.

### What exists today (grounding)
- **Throughput harness** `benches/trainer_throughput.rs` (criterion, `harness = false`, registered as `[[bench]] name = "trainer_throughput"` in `Cargo.toml`). It defines four groups:
  - `a2c_train_step`, `ppo_train_step` — isolated single gradient step on a fixed seeded synthetic batch (`SYNTH_BATCH = 64`), reported via `Throughput::Elements(SYNTH_BATCH)`.
  - `a2c_cartpole_steps_per_sec`, `ppo_cartpole_steps_per_sec` — full rollout-collect (16 envs x 5 steps CartPole) + one update, reported via `Throughput::Elements(n_steps * num_envs)` so criterion prints env-steps/sec.
  - Conventions: CPU backend `type B = Autodiff<NdArray<f32>>`; fresh seeded trainer per `iter_batched` iteration; `BatchSize::SmallInput`; `black_box` the returned stats.
- **Learning-curve CSV**: the A2C (`examples/games/cartpole/train_cartpole_a2c.rs`) and PPO (`train_cartpole_modern.rs`) CartPole examples emit `env_steps,mean_episode_reward` rows via the opt-in `CURVE_CSV` env var, gated through an `open_curve_csv()` helper that writes the header and returns `Ok(None)` when unset (zero behavior change). Seeded policy init makes re-runs reproducible.
- **DQN**: trainer `src/train/dqn/` (`DQNTrainerBurn`, off-policy replay buffer, `train_step(&mut rng, forward_fn, forward_target_fn) -> Result<Option<DQNStepStatsBurn>>` returns `None` until `min_buffer_size` warmup is reached). Example `examples/games/cartpole/train_cartpole_dqn.rs` on CartPole (discrete, **same env as PPO/A2C** so directly comparable). No `CURVE_CSV` yet.
- **SAC**: trainer `src/train/sac/` (`SacTrainer::<B>::new(config, obs_dim, action_dim, device)`, off-policy, `train() -> Result<Option<SacStepStats>>`). Env is **continuous** `PendulumSwingUp` (`src/env/games/pendulum.rs`, `with_seed`, action rescaled by `MAX_TORQUE = 2.0`) — NOT CartPole, so SAC benchmarks form a separate Pendulum group. Convergence test exists at `tests/test_sac_pendulum.rs`. **No `train_sac` example exists yet** — creating it is part of this work.

### Architectural decisions
See the pinned design comment below for the full rationale. Summary:

1. **Throughput** — add four groups matching existing conventions:
   - `dqn_train_step` — isolated single Double-DQN gradient step. Because `train_step` samples from the replay buffer and no-ops until warmup, the bench fixture **pre-fills the buffer to `min_buffer_size`** with seeded synthetic CartPole transitions in the `iter_batched` setup closure (untimed), so every timed iteration performs exactly one real gradient step. Reported via `Throughput::Elements(batch_size)` to mirror `a2c_train_step`/`ppo_train_step` (per-update cost).
   - `dqn_cartpole_steps_per_sec` — full env-loop: step N seeded CartPole envs, push to buffer, one gradient update per env step (matching the example's loop). Reported via `Throughput::Elements(env_steps)`. **Buffer warmup is pre-filled untimed in setup** so the timed region measures steady-state off-policy throughput, not the warmup transient.
   - `sac_train_step` — isolated single SAC update on a Pendulum replay buffer pre-filled to `min_buffer_size` (untimed setup). Reported via `Throughput::Elements(batch_size)`.
   - `sac_pendulum_steps_per_sec` — full Pendulum env-loop steady-state, `Throughput::Elements(env_steps)`.

2. **Off-policy vs on-policy measurement fairness (the honest comparison)** — documented explicitly in the bench module header:
   - The `*_train_step` groups ARE directly comparable across all four algorithms: each is "cost of one gradient update on a fixed-size batch." This is the apples-to-apples per-update number.
   - The `*_steps_per_sec` groups are **NOT** directly comparable on-policy vs off-policy and must be labeled as such. On-policy steps/sec = (rollout collect of `n_steps*num_envs` + one update) amortized; off-policy steps/sec = (one env step + one gradient update + replay sample) per env step. They measure different work per env step. The fairness convention: **pre-fill replay warmup untimed** so off-policy numbers reflect steady-state, and **document that on-policy does ~`batch` env-interactions per update while off-policy does 1 env-step per update** — so steps/sec compares within an algorithm class (A2C vs PPO; DQN vs SAC), not across classes. The header will state this caveat verbatim.

3. **Learning curves** — `CURVE_CSV` emission with the same `env_steps,mean_episode_reward` schema and `open_curve_csv()` helper:
   - DQN CartPole example: add the helper + per-log-interval row emission. CartPole reward is +1/step so the curve overlays directly on A2C/PPO curves (same env, seed, budget).
   - SAC: create a new seeded `train_sac` Pendulum example (does not exist), then add `CURVE_CSV` emission. Pendulum returns are negative (random ~ -1200..-1600, near-optimal ~ -150), so SAC curves form a separate Pendulum comparison group, not overlaid on CartPole.

### Decomposition (builder-tractable, ordered)
- **PR 1 (DQN bench + curve)** — additive to `benches/trainer_throughput.rs` and `examples/games/cartpole/train_cartpole_dqn.rs`. No new files, same env as existing CartPole benches. Lowest blast radius. **Immediately buildable.**
- **PR 2 (SAC Pendulum example + curve)** — new example file `examples/games/pendulum/train_sac.rs` + `[[example]]` registration. Establishes the seeded Pendulum training loop the SAC bench reuses. Independent of PR 1. **Immediately buildable.**
- **PR 3 (SAC bench groups)** — adds `sac_train_step` + `sac_pendulum_steps_per_sec` to the harness. **Depends on PR 2** (reuses the seeded Pendulum loop conventions / scale_action helper established there) so the two SAC pieces stay consistent.

### Blast radius
All changes are additive: new bench groups (registered in the existing `criterion_group!`), `CURVE_CSV`-gated emission (no-op when unset), and one new example. No trainer/library API changes. No behavior change to existing examples or benches when the new env vars are unset.

Child issues and DAG are listed in a follow-up comment.

**Comment by rjwalters** [comment]
## Architectural design (pinned)

### The off-policy vs on-policy fairness problem, resolved

The harness has two kinds of group. They demand different fairness treatment:

**`*_train_step` (per-update cost) — fully comparable across all 4 algos.**
Each measures "one gradient update on a fixed batch." For A2C/PPO the batch is the seeded `synthetic_batch` (`SYNTH_BATCH = 64`). For DQN/SAC the batch comes from a replay buffer, and `train_step`/`train()` return `None` until `min_buffer_size` is reached. Resolution: **pre-fill the replay buffer to `min_buffer_size` in the `iter_batched` setup closure (untimed)** with seeded synthetic transitions, so every timed iteration performs exactly one real gradient step. Report `Throughput::Elements(batch_size)`. This is the honest cross-algorithm per-update number.

- DQN: `train_step(&mut rng, |q,o| q.forward(o), |q,o| q.forward(o))` — fixture builds `QNetworkBurn`, `DQNTrainerBurn`, pushes `>= min_buffer_size` seeded `(obs, action, reward, next_obs, done)` CartPole-shaped transitions.
- SAC: `trainer.train()` — fixture builds `SacTrainer::<B>::new(cfg, 3, 1, device)` (Pendulum obs_dim=3, action_dim=1), pushes `>= min_buffer_size` seeded transitions, and ensures `total_env_steps >= learning_starts` via `increment_env_step()`.

**`*_steps_per_sec` (full-loop env throughput) — comparable WITHIN an algo class only.**
The dishonest comparison would be "DQN does 60k env steps/sec vs A2C does 200k env steps/sec, therefore A2C is faster." That conflates different work-per-env-step:
- On-policy (A2C/PPO): collect `n_steps*num_envs = 80` env interactions, then ONE update amortized over those 80 steps.
- Off-policy (DQN/SAC): ONE env step + ONE replay-sampled gradient update + buffer push, every single env step.

So off-policy does ~`batch_size` more gradient work per env step than on-policy does. The harness will state this caveat **verbatim in the module header** and label the groups so readers compare A2C-vs-PPO and DQN-vs-SAC, with `*_train_step` as the only cross-class number. Additionally, **replay warmup is pre-filled untimed** in the steps/sec setup closure so off-policy numbers reflect steady-state throughput (one update per step) rather than the warmup transient (env steps with no update).

### Reproducibility
Every fixture uses a fixed seed: trainer config `.seed(...)`, `StdRng::seed_from_u64(...)` for action/replay sampling, and `PendulumSwingUp::with_seed(...)`. Matches the existing seeded A2C/PPO fixtures.

### Learning-curve schema
Unchanged `env_steps,mean_episode_reward` two-column CSV with the `open_curve_csv()` helper copied verbatim (writes header, returns `Ok(None)` when `CURVE_CSV` unset). DQN/SAC examples emit one row per existing log interval. DQN on CartPole overlays directly on A2C/PPO (same env/seed/budget); SAC on Pendulum is a separate negative-return curve group.

**Comment by rjwalters** [comment]
## Child issues + DAG

| Issue | Title | Labels | Depends on | Buildable now |
|-------|-------|--------|------------|---------------|
| #160 | DQN benchmark coverage: throughput groups + CURVE_CSV (CartPole) | `loom:issue`, `tier:goal-advancing` | — | **Yes** |
| #162 | SAC Pendulum example: seeded train_sac + CURVE_CSV (new example) | `loom:issue`, `tier:goal-advancing` | — | **Yes** |
| #165 | SAC benchmark coverage: sac_train_step + sac_pendulum_steps_per_sec | `loom:blocked`, `tier:goal-advancing` | #162 | No (until #162 merges) |

### DAG
```
#160 (DQN bench + curve) ───────────────► (independent, leaf)

#162 (SAC example + curve) ──────────────► #165 (SAC bench groups)
```

- #160 and #162 are independent and immediately buildable (both `loom:issue`).
- #165 is `loom:blocked` on #162 (reuses the seeded Pendulum loop / `scale_action` helper). When #162 merges, relabel #165 `loom:issue`.

All work is additive: new bench groups, `CURVE_CSV`-gated emission (no-op when unset), one new example. No trainer/library API changes.

**Comment by rjwalters** [comment]
## Benchmark-extension epic complete ✅
- #160 (PR #173): DQN throughput groups + CURVE_CSV on the DQN CartPole example
- #162 (PR #172): SAC Pendulum training example + CURVE_CSV
- #165 (PR #175): SAC throughput groups

The criterion harness `benches/trainer_throughput.rs` now covers all four single-agent algorithms (PPO, A2C, DQN, SAC) with per-update and full-loop steps/sec, plus a documented on-policy vs off-policy fairness caveat. Learning-curve CSV (`CURVE_CSV`) is available on the A2C, PPO, DQN, and SAC examples. Closing.
